### DISCUSSION TITLE
Throw expression disallowed in ref ternary

### DISCUSSION THREAD

**Discussion by jcouv** [dotnet/csharplang#919] [General]
`i = b ? ref x : throw null;` gives a surprising `error CS8402: Both conditional operator values must be ref values or neither may be a ref value`

This is disallowed in 7.2 (which introduces ref ternaries), but maybe it should be allowed later on (it seems irregular not to allow).

Should the syntax be:
- `i = b ? ref x : throw null; `, or
- `i = b ? ref x : ref throw null;`?

FYI @VSadov @gafter

**Comment by svick** [comment]
```c#
i = b ? ref x : ref throw null;
```

I don't think this syntax makes sense, there is no reference to anything in the else expression, so there shouldn't be a `ref`.

**Comment by VSadov** [comment]
I like the case where `throw` does not need ref .

We do not require `ref` in arrow bodies when they throw.  (since they do not really return anything)

```cs
     public ref readonly C M2() => throw null;
```

**Comment by DavidArno** [comment]
I'm with @svick here in thinking that `ref throw` makes no sense. `throw` has no type, so "ref of no type" seems nonsensical to me. Plus I feel it's ungainly syntax. So just `i = b ? ref x : throw null;` ought to do the job.

**Comment by sharwell** [comment]
This appears to be a bug in Roslyn's implementation of C# 7.2. Is there a reason it's filed here?

While [conditional-ref.md](https://github.com/dotnet/csharplang/blob/master/proposals/conditional-ref.md) doesn't call out this case specifically, the [throw-expression.md](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.0/throw-expression.md) document is clear about the behavior:

> A *throw_expression* is convertible to every type by an implicit conversion.

Even if `ref` types are "special", the intent of the statement for throw expressions is also clear. With no negative impact on verification or safety, plus the necessity of throwing exceptions for situations described in issues like #497, I can't see any reason why someone would object to this particular form.

**Comment by VSadov** [comment]
Not technically a bug. There are very few places where throw expression is allowed - cant assign or pass as a method argument for example, ref or not.  

It seems everyone agrees that ref ternary should allow it.

**Comment by sharwell** [comment]
> There are very few places where throw expression is allowed - cant assign or pass as a method argument for example

True, but the syntax in question is one of the three places where it is explicitly allowed.

**Comment by xparadoxical** [comment]
The 2nd example currently produces `error CS8115: A throw expression is not allowed in this context.`. Why was it decided to disallow this?

**Reply by gafter** [reply]
The fact that it is disallowed simply follows from the specification of the individual features involved (a ref ternary expression requires ref on both the consequence and the alternative; a ref expression requires a variable as its target; a throw expression does not designate a variable).  As you can see from the discussion, above, it seems like the C# team is open to extending the language to allow it.
